### PR TITLE
[8.0][ADD] BI report for the business requirements

### PR DESCRIPTION
**Pull request by serpentcs-dev1**
@elicoidal I have added the BI report for the business requirement.

Can you please review?

cc @sudhir-serpentcs

### CODE REVIEW

**Comment by martintorresvm** [comment]
@sudhir-serpentcs Fix Travis please

cc @elicoidal

**Comment by JayVora-SerpentCS** [comment]
@darshan-serpent Please do the needful for the travis and runbot.

**Comment by sudhir-serpentcs** [comment]
@darshan-serpent Please fix the travis.
CC: @victormartinelicocorp

**Comment by elicoidal** [comment]
@darshan-serpent @JayVora-SerpentCS we do not have Enterprise CLA for Serpent consulting covering all your team members. Would it be a good idea to have it?
Otherwise please send an Individual CLA for Yogesh

**Comment by elicoidal** [comment]
@gurneyalex I have the following error for runbot:
> Internal Server Error
> The server encountered an internal error and was unable to complete your request. Either the server is overloaded or there is an error in the application.

File: business_requirement_deliverable/report/br_deliverable_report.py
Comment (elicoidal) [inline business_requirement_deliverable/report/br_deliverable_report.py:17]: We should have both Master project and Linked project

File: business_requirement_deliverable/report/br_deliverable_report.py
Comment (elicoidal) [inline business_requirement_deliverable/report/br_deliverable_report.py]: better this order:
    dlv_description = fields.Text('Deliverable Description', readonly=True)
    dlv_product = fields.Many2one('product.product', 'Dlv Product',
                                  readonly=True)
    res_description = fields.Text('Resource Description', readonly=True)
    res_product = fields.Many2one('product.product', 'Res Product',
                                  readonly=True)

File: business_requirement_deliverable/report/br_deliverable_sale_report.py
Comment (elicoidal) [inline business_requirement_deliverable/report/br_deliverable_sale_report.py]: Change the order

File: business_requirement_deliverable_cost/report/br_deliverable_cost_report.py
Comment (elicoidal) [inline business_requirement_deliverable_cost/report/br_deliverable_cost_report.py]: order

File: business_requirement_deliverable/security/ir.model.access.csv
Comment (elicoidal) [inline business_requirement_deliverable/security/ir.model.access.csv]: about the view access, I think it should be available for both with same rights. It does not make sense to have 1,1,1,1 right?

Comment (elicoidal) [reply business_requirement_deliverable/security/ir.model.access.csv]: This is not correct because there is current a flaw in the design of the ACL. Please check https://github.com/OCA/business-requirement/issues/91.
This needs to be fixed and then the BI report with sales should only be available for the BR estimation group.

File: business_requirement_deliverable_cost/security/ir.model.access.csv
Comment (elicoidal) [inline business_requirement_deliverable_cost/security/ir.model.access.csv]: only available to cost control group "group_business_requirement_cost_control"

**Review by elicoidal — CHANGES_REQUESTED** [review]
We need to fix the issue https://github.com/OCA/business-requirement/issues/91 first and finetune the ACL for this module after that
Other than that :+1: 
Not yet tested in runbot

File: business_requirement_deliverable/report/br_deliverable_report.py
Comment (dreispt) [inline business_requirement_deliverable/report/br_deliverable_report.py]: For consistency, please name it as `business.requirement.deliverable.report`

File: business_requirement_deliverable/report/br_deliverable_sale_report.py
Comment (dreispt) [inline business_requirement_deliverable/report/br_deliverable_sale_report.py]: `business.requirement.deliverable.sale.report`

File: business_requirement_deliverable/report/br_deliverable_sale_report.py
Comment (dreispt) [inline business_requirement_deliverable/report/br_deliverable_sale_report.py]: `business_requirement_deliverable_sale_report`

File: business_requirement_deliverable_cost/models/business.py
Comment (dreispt) [inline business_requirement_deliverable_cost/models/business.py]: The indentation was correct

File: business_requirement_deliverable_cost/report/br_deliverable_cost_report.py
Comment (dreispt) [inline business_requirement_deliverable_cost/report/br_deliverable_cost_report.py]: Prepend `business.requirement.`

File: business_requirement_deliverable_cost/report/br_deliverable_cost_report.py
Comment (dreispt) [inline business_requirement_deliverable_cost/report/br_deliverable_cost_report.py]: Prepend `business_requirement_`

**Comment by dreispt** [comment]
Make Mr.Travis happy and we're good to go!

**Comment by elicoidal** [comment]
pending this: https://github.com/OCA/business-requirement/pull/94#pullrequestreview-28811664

**Comment by serpentcs-dev1** [comment]
Fixed Travis

@elicoidal @victormartinelicocorp @sudhir-serpentcs

File: business_requirement_deliverable/report/br_deliverable_report.py
Comment (martintorresvm) [reply business_requirement_deliverable/report/br_deliverable_report.py:17]: @darshan-serpent any reason you dismiss this request?
Could you add `linked_project` field after project_id?

**Comment by elicoidal** [comment]
Thanks we have to wait for https://github.com/OCA/business-requirement/pull/86#issuecomment-288903426

File: business_requirement_deliverable/report/br_deliverable_report.py
Comment (serpentcs-dev1) [reply business_requirement_deliverable/report/br_deliverable_report.py:17]: @victormartinelicocorp 
due to linked_project field is in 'business_requirement_deliverable_project' module so can't access it in 'business_requirement_deliverable' module. 
So what can I do like dependency or any other solution?

Comment (martintorresvm) [reply business_requirement_deliverable/report/br_deliverable_report.py:17]: @elicoidal Not sure about what we should do

IMO Add dependency to business_requirement_deliverable_project is not an option so we could choose:

1) Dismiss to add `linked_project` field to the report
2) Create the BI report on `business_requirement_deliverable_project` instead of `business_requirement_deliverable`
3) Refactor `business_requirement_deliverable_project` to move field `linked_project` to `business_requirement_deliverable`

cc @darshan-serpent @sudhir-serpentcs

**Comment by serpentcs-dev1** [comment]
Added BI report file and improved code.

@elicoidal @victormartinelicocorp 
CC @sudhir-serpentcs

File: business_requirement_deliverable/report/br_deliverable_report.py
Comment (elicoidal) [reply business_requirement_deliverable/report/br_deliverable_report.py:17]: option 1.

**Comment by elicoidal** [comment]
@darshan-serpent Wait for #100 to be merged then rebase/fix xml_id

**Comment by elicoidal** [comment]
@darshan-serpent fix #100 is merged. Can you rebase and fix the ACL groups so that we can merge this one as well?

**Comment by serpentcs-dev1** [comment]
Solved conflict.

@elicoidal  @victormartinelicocorp 
CC @sudhir-serpentcs

File: business_requirement_deliverable/models/business.py
Comment (elicoidal) [inline business_requirement_deliverable/models/business.py]: Somewhere the information was not properly transmitted.
I said that we dont need this in the report. So no need to transfer the field to brd module

File: business_requirement_deliverable/report/br_deliverable_sale_report_view.xml
Comment (elicoidal) [inline business_requirement_deliverable/report/br_deliverable_sale_report_view.xml]: string: Sales Analysis

File: business_requirement_deliverable/report/br_deliverable_sale_report_view.xml
Comment (elicoidal) [inline business_requirement_deliverable/report/br_deliverable_sale_report_view.xml:22]: This sales report should only be available for the group "estimation"

File: business_requirement_deliverable_cost/report/br_deliverable_cost_report_view.xml
Comment (elicoidal) [inline business_requirement_deliverable_cost/report/br_deliverable_cost_report_view.xml]: s/cost/Cost

File: business_requirement_deliverable_cost/report/br_deliverable_cost_report_view.xml
Comment (elicoidal) [inline business_requirement_deliverable_cost/report/br_deliverable_cost_report_view.xml:25]: This menu should only available to the "cost" group

File: business_requirement_deliverable_project/models/business.py
Comment (elicoidal) [inline business_requirement_deliverable_project/models/business.py]: You need to undo this.

File: business_requirement_deliverable_project/models/project.py
Comment (elicoidal) [inline business_requirement_deliverable_project/models/project.py]: irrelevant to this BR

File: business_requirement_deliverable_cost/models/business.py
Comment (elicoidal) [reply business_requirement_deliverable_cost/models/business.py]: indeed irrelevant to this BR

File: business_requirement_deliverable_project/report/__init__.py
Comment (elicoidal) [inline business_requirement_deliverable_project/report/__init__.py]: This report should be in business_requirement_deliverable, not in this module.

**Review by elicoidal — CHANGES_REQUESTED** [review]
Many comments:
* 2 reports should in brd modules
* 1 report in brd_cost module
ACL must be fixed

**Comment by serpentcs-dev1** [comment]
@elicoidal @victormartinelicocorp 
Please review code.

CC @sudhir-serpentcs

File: business_requirement_deliverable_cost/report/br_deliverable_cost_report.py
Comment (elicoidal) [inline business_requirement_deliverable_cost/report/br_deliverable_cost_report.py:42]: :+1: 
Functional tests are fine but I have 2 small additional requests:
* Can you add Gross Profit field? It is Total revenue - Total cost
* Can add the description of the BR (after name)

**Review by elicoidal — CHANGES_REQUESTED** [review]

File: business_requirement_deliverable/report/br_deliverable_sale_report_view.xml
Comment (martintorresvm) [inline business_requirement_deliverable/report/br_deliverable_sale_report_view.xml]: @darshan-serpent Please help reviewers. XML parameters should be place on one column, like this:

action="action_deliverable_sale_report" 
id="menu_action_deliverable_sale_report" 
parent="base.menu_project_report"
groups="business_requirement_deliverable.group_business_requirement_estimation"

File: business_requirement_deliverable_cost/report/br_deliverable_cost_report.py
Comment (martintorresvm) [inline business_requirement_deliverable_cost/report/br_deliverable_cost_report.py]: @darshan-serpent Please review and change all fields parameters for models to layout indented as column, like this:
'project.project',
'Master Project',
readonly=True)

**Review by martintorresvm — CHANGES_REQUESTED** [review]

**Comment by serpentcs-dev1** [comment]
@elicoidal @victormartinelicocorp 
please review code.

CC @sudhir-serpentcs

**Review by elicoidal — APPROVED** [review]
:+1:

**Comment by elicoidal** [comment]
@pedrobaeza @dreispt I think this one is good to go.

**Comment by elicoidal** [comment]
@seb-elico

File: business_requirement_deliverable/report/br_deliverable_report.py
Comment (seb-elico) [inline business_requirement_deliverable/report/br_deliverable_report.py]: Please improve SQL readability. I suggest:
````
            CREATE VIEW business_requirement_deliverable_report AS (
                SELECT
                    br.id,
                    br.name,
                    br.description,
                    br.partner_id,
                    br.project_id,
                    br.change_request,
                    br.priority,
                    dlv.product_id AS dlv_product,
                    dlv.name AS dlv_description,
                    res.product_id AS res_product,
                    res.name AS res_description,
                    count(DISTINCT br.id) AS br_count,
                    count(DISTINCT dlv.id) AS dlv_count,
                    count(DISTINCT res.id) AS res_count,
                    res.qty AS res_qty,
                    dlv.qty AS dlv_qty
                FROM
                    business_requirement br
                    FULL OUTER JOIN business_requirement_deliverable dlv
                        ON br.id = dlv.business_requirement_id
                    FULL OUTER JOIN business_requirement_resource res
                        ON res.business_requirement_deliverable_id = dlv.id
                GROUP BY dlv.id, br.id, res.id
            )
````

File: business_requirement_deliverable/report/br_deliverable_sale_report.py
Comment (seb-elico) [inline business_requirement_deliverable/report/br_deliverable_sale_report.py]: Please improve SQL readability (following previous suggestion)

File: business_requirement_deliverable_cost/report/br_deliverable_cost_report.py
Comment (seb-elico) [inline business_requirement_deliverable_cost/report/br_deliverable_cost_report.py]: Please improve SQL readability

**Review by seb-elico — CHANGES_REQUESTED** [review]

File: business_requirement_deliverable_cost/report/br_deliverable_cost_report.py
Comment (pedrobaeza) [reply business_requirement_deliverable_cost/report/br_deliverable_cost_report.py]: Better to decompose SQL creation in submethods for improving inheritance, like in https://github.com/odoo/odoo/blob/9e159ef2048574d179a9afb2226397e962aa5725/addons/account/report/account_invoice_report.py#L163-L180

**Comment by serpentcs-dev1** [comment]
@seb-elico @pedrobaeza 
please review code

cc @sudhir-serpentcs @elicoidal

File: business_requirement_deliverable/report/br_deliverable_report.py
Comment (pedrobaeza) [inline business_requirement_deliverable/report/br_deliverable_report.py]: Incorrect indentation

File: business_requirement_deliverable/report/br_deliverable_report.py
Comment (pedrobaeza) [inline business_requirement_deliverable/report/br_deliverable_report.py]: Incorrect indentation

File: business_requirement_deliverable/report/br_deliverable_report.py
Comment (pedrobaeza) [inline business_requirement_deliverable/report/br_deliverable_report.py]: Incorrect indentation

**Comment by serpentcs-dev1** [comment]
@seb-elico @pedrobaeza
Improved Indentation
Please review code

cc @sudhir-serpentcs @elicoidal

**Review by pedrobaeza — APPROVED** [review]
Please squash on merge.

**Comment by elicoidal** [comment]
Thanks @pedrobaeza  for the review!

**Pull request merged**
